cmd/zrepl/mgmt_conn.c

```diff
@@ -0,0 +1,899 @@
+				(c)->conn_host, (c)->conn_port, ##__VA_ARGS__)
+#else
+#define	DBGCONN(c, fmt, ...)
+#endif
+
+/* Max # of events from epoll processed at once */
+#define	MAX_EVENTS	10
+#define	MGMT_PORT	"12000"
+#define	RECONNECT_DELAY	5	// 5 seconds
+
+/*
+ * Mgmt connection states.
+ */
+enum conn_state {
+	CS_CONNECT,		// tcp connect is in progress
+	CS_INIT,		// initial state or state after sending reply
+	CS_READ_VERSION,	// reading request version
+	CS_READ_HEADER,		// reading request header
+	CS_READ_PAYLOAD,	// reading request payload
+	CS_CLOSE,		// closing connection - final state
+};
+
+/*
+ * Structure representing mgmt connection and all its reading/writing state.
+ */
+typedef struct uzfs_mgmt_conn {
+	SLIST_ENTRY(uzfs_mgmt_conn) conn_next;
+	int		conn_fd;	// network socket FD
+	int		conn_refcount;	// should be 0 or 1
+	char		conn_host[MAX_IP_LEN];
+	uint16_t	conn_port;
+	enum conn_state	conn_state;
+	void		*conn_buf;	// buffer to hold network data
+	int		conn_bufsiz;    // bytes to read/write in total
+	int		conn_procn;	// bytes already read/written
+	zvol_io_hdr_t	*conn_hdr;	// header of currently processed cmd
+	time_t		conn_last_connect;  // time of last attempted connect()
+} uzfs_mgmt_conn_t;
+
+/* conn list can be traversed or changed only when holding the mutex */
+kmutex_t	conn_list_mtx;
+SLIST_HEAD(, uzfs_mgmt_conn) uzfs_mgmt_conns;
+
+/* event FD for waking up event loop thread blocked in epoll_wait */
+int mgmt_eventfd;
+int epollfd = -1;
+/* default iSCSI target IP address */
+char *target_addr;
+
+static int move_to_next_state(uzfs_mgmt_conn_t *conn);
+
+/*
+ * Remove connection FD from poll set and close the FD.
+ */
+static int
+close_conn(uzfs_mgmt_conn_t *conn)
+{
+	/* Release resources tight to the conn */
+	if (conn->conn_buf != NULL) {
+		kmem_free(conn->conn_buf, conn->conn_bufsiz);
+		conn->conn_buf = NULL;
+	}
+	conn->conn_bufsiz = 0;
+	conn->conn_procn = 0;
+	if (conn->conn_hdr != NULL) {
+		kmem_free(conn->conn_hdr, sizeof (zvol_io_hdr_t));
+		conn->conn_hdr = NULL;
+	}
+
+	if (epoll_ctl(epollfd, EPOLL_CTL_DEL, conn->conn_fd, NULL) == -1) {
+		perror("epoll_ctl del");
+		return (-1);
+	}
+	(void) close(conn->conn_fd);
+	conn->conn_fd = -1;
+	return (0);
+}
+
+/*
+ * Create non-blocking socket and connect to the target.
+ * Returns the new FD or -1.
+ */
+static int
+connect_to_tgt(uzfs_mgmt_conn_t *conn)
+{
+	struct sockaddr_in istgt_addr;
+	int sfd, rc;
+
+	conn->conn_last_connect = time(NULL);
+
+	bzero((char *)&istgt_addr, sizeof (istgt_addr));
+	istgt_addr.sin_family = AF_INET;
+	istgt_addr.sin_addr.s_addr = inet_addr(conn->conn_host);
+	istgt_addr.sin_port = htons(conn->conn_port);
+
+	sfd = create_and_bind(MGMT_PORT, B_FALSE, B_TRUE);
+	if (sfd < 0)
+		return (-1);
+
+	rc = connect(sfd, (struct sockaddr *)&istgt_addr, sizeof (istgt_addr));
+	if (rc != 0 && errno != EINPROGRESS) {
+		close(sfd);
+		DBGCONN(conn, "Failed to connect");
+		return (-1);
+	}
+	return (sfd);
```

Comment:
need to handle EINPROGRESS

Reply:
or does this mean - we get an EPOLLOUT event on successful connection?

Reply:
yes. EINPROGRESS is either followed by EPOLLOUT (success) or EPOLLERR (failure). I will add a comment explaining that.

Reply:
@jkryl actually, I meant - for successful connection here with CONNECT call returning 0, we are assigning the state as CS_CONNECT in the caller. And, it seems - mgmt_thread is expecting EPOLLOUT on this.
Same path goes for EINPROGRESS case also. That is bit unclear for me.
And, many thanks for adding comment

Reply:
Hi Vitta, in reality it's unlikely to happen that connect on non-blocking socket would return 0, because connect cannot be done in a non-blocking way by nature. However even if that would happen there is no harm. Connected socket will return EPOLLOUT (signal that the socket can be written) and that will immediately transition the state from CS_CONNECT to CS_INIT (the state transition path is the same as in EINPROGRESS case).

Reply:
thanks @jkryl for explaining